File: src/sender/sender.cs
// xywdl_sender — 校园网认证请求发送器 (C# 备用发送层)
//
// 用途:
//   xywdl.ps1 用 Invoke-WebRequest 发送失败时的第二层保底。
//   从 stdin 读入完整 quickauth.do URL (已 URL 编码), 发 HTTP GET,
//   把响应体原样输出到 stdout, 由主脚本判定认证结果。
//
// 语义:
//   exit 0   = 成功发出请求并拿到响应体 (stdout 为响应体, 即使 4xx/5xx)
//   exit 1   = 网络层失败 (连接失败 / 超时 / DNS 失败等)
//   exit 2   = 参数错误 (没有收到 URL)
//
// 特性:
//   - 仅依赖 .NET Framework 4.x (Windows 7+ 自带, 无需额外运行时)
//   - Proxy = null: 直连, 避免被系统代理 / 校园网网关拦截
//   - AllowAutoRedirect = false: 不跟随 302, 与 PS 端 -MaximumRedirection 0 一致
//   - 超时 15 秒, 防止卡死
//
// 构建:
//   "%SystemRoot%\Microsoft.NET\Framework64\v4.0.30319\csc.exe" /nologo /optimize+ /target:exe /out:xywdl_sender.exe sender.cs
//
// 用法 (由脚本调用, 不直接给用户用):
//   echo "http://.../quickauth.do?userid=..." | xywdl_sender.exe
//   或: xywdl_sender.exe "http://.../quickauth.do?userid=..."

using System;
using System.IO;
using System.Net;
using System.Text;

class XywdlSender
{
    static int Main(string[] args)
    {
        string url;
        if (args.Length > 0)
        {
            url = args[0].Trim();
        }
        else
        {
            // 从 stdin 读完整 URL (避免明文密码出现在进程命令行里)
            url = Console.In.ReadToEnd().Trim();
        }

        if (string.IsNullOrEmpty(url))
        {
            Console.Error.WriteLine("[sender] 未提供 URL");
            return 2;
        }

        // 剥掉可能的 BOM (U+FEFF): PS 管道用带 BOM 的 UTF8 编码传字符串时,
        // 原生进程会收到开头的 BOM 字符, 不剥掉会导致 WebRequest 报"URI 方案无效"
        url = url.Trim().TrimStart('\uFEFF');

        try
        {
            HttpWebRequest req = (HttpWebRequest)WebRequest.Create(url);
            req.Method = "GET";
            req.Proxy = null;                 // 直连, 绕开系统代理
            req.Timeout = 15000;              // 连接超时 (毫秒)
            req.ReadWriteTimeout = 15000;     // 读响应超时 (毫秒)
            req.AllowAutoRedirect = false;    // 不跟随重定向, 看原始响应
            req.UserAgent = "XXGCXY-CampusNet-AutoLogin/2.0";
            req.Accept = "text/html,application/json,*/*";

            HttpWebResponse resp = (HttpWebResponse)req.GetResponse();
            using (resp)
            {
                using (StreamReader sr = new StreamReader(resp.GetResponseStream(), Encoding.UTF8))
                {
                    string body = sr.ReadToEnd();
                    Console.Write(body);
                }
                return 0;
            }
        }
        catch (WebException we)
        {
            // 4xx/5xx (甚至 3xx 被禁后) 会走这里, 但能拿到响应对象就读取 body,
            // 交由主脚本判定; 只有连响应对象都拿不到才是网络层失败。
            HttpWebResponse resp = we.Response as HttpWebResponse;
            if (resp != null)
            {
                using (resp)
                using (StreamReader sr = new StreamReader(resp.GetResponseStream(), Encoding.UTF8))
                {
                    string body = sr.ReadToEnd();
                    Console.Write(body);
                }
                return 0;
            }
            Console.Error.WriteLine("[sender] 请求失败: " + we.Message);
            return 1;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("[sender] 错误: " + ex.Message);
            return 1;
        }
    }
}

File: src/sender/sender.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
xywdl_sender — 校园网认证请求发送器 (Python 保底层, 跨平台)
============================================================

用途:
  xywdl.ps1 / xywdl.sh 的"最强保底"发送层。当 PowerShell Invoke-WebRequest
  或 C# sender 都不可用时, 用系统 python3 发送认证请求。

  - 跨平台: Windows / Linux / macOS 都可用 (Python 3.6+)
  - 零依赖: 只用标准库 urllib, 不需要 pip 安装任何包
  - 语义与 C# sender 完全一致, 主脚本可共用一套判定逻辑

语义:
  exit 0   = 成功发出请求并拿到响应体 (stdout 为响应体, 即使 4xx/5xx)
  exit 1   = 网络层失败 (连接失败 / 超时 / DNS 失败等)
  exit 2   = 参数错误 (没有收到 URL)

用法 (由脚本调用, 不直接给用户用):
  echo "http://.../quickauth.do?userid=..." | python3 sender.py
  或: python3 sender.py "http://.../quickauth.do?userid=..."
"""

import sys

if sys.version_info < (3, 6):
    sys.stderr.write("[sender] 需要 Python 3.6+\n")
    sys.exit(1)


def _read_stdin_url() -> str:
    """从 stdin 读完整 URL。

    用原始字节读再按 utf-8-sig 解码: 同时解决两个问题——
      - PS 管道用带 BOM 的 UTF8 编码传字符串, utf-8-sig 会把 BOM 剥掉
      - 避免文本模式按 locale 编码误读中文/特殊字符
    """
    raw = sys.stdin.buffer.read()
    try:
        s = raw.decode("utf-8-sig")
    except UnicodeDecodeError:
        s = raw.decode("latin-1")
    # 兜底: 剥掉所有开头的 BOM (极端情况可能叠多个)
    return s.lstrip("\ufeff").strip()


def decode_body(raw: bytes, content_type: str = "") -> str:
    """优先按 header 里的 charset 解码, 其次 UTF-8, 最后 Latin-1 兜底。"""
    charset = None
    if content_type:
        # 从 "text/html; charset=gbk" 之类里抠出 charset
        for part in content_type.split(";"):
            part = part.strip().lower()
            if part.startswith("charset="):
                charset = part.split("=", 1)[1].strip().strip('"')
                break
    for enc in (charset, "utf-8", "gbk", "latin-1"):
        if not enc:
            continue
        try:
            return raw.decode(enc)
        except (UnicodeDecodeError, LookupError):
            continue
    return raw.decode("utf-8", errors="replace")


def main() -> int:
    url = (sys.argv[1].strip() if len(sys.argv) > 1
           else _read_stdin_url())
    if not url:
        sys.stderr.write("[sender] 未提供 URL\n")
        return 2

    try:
        import urllib.request
        import urllib.error

        # 禁用代理, 直连 (与 PS -Proxy $null / C# req.Proxy=null 一致)
        opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({}),
        )
        req = urllib.request.Request(url, headers={
            "User-Agent": "XXGCXY-CampusNet-AutoLogin/2.0",
            "Accept": "text/html,application/json,*/*",
        })

        try:
            with opener.open(req, timeout=15) as resp:
                raw = resp.read()
                content_type = resp.headers.get("Content-Type", "")
                sys.stdout.write(decode_body(raw, content_type))
                return 0
        except urllib.error.HTTPError as e:
            # 4xx/5xx: 拿 body 交给主脚本判定, 算"发出去了"
            raw = e.read()
            content_type = e.headers.get("Content-Type", "")
            sys.stdout.write(decode_body(raw, content_type))
            return 0
    except Exception as ex:  # noqa: BLE001 - 保底层要兜住一切异常
        sys.stderr.write("[sender] 请求失败: %s\n" % ex)
        return 1


if __name__ == "__main__":
    sys.exit(main())
